### PR TITLE
Switch to capstone 5.0

### PR DESCRIPTION
**Pull request by wargio**
# DO NOT SQUASH!!!

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

I have hand-verified all the instructions (encoding/decoding) using objdump

### CODE REVIEW

File: meson_options.txt
Comment (ret2libc) [inline meson_options.txt]: Shouldn't we rename it v5 instead?

Comment (wargio) [reply meson_options.txt]: yes. we should.

**Review by XVilka — CHANGES_REQUESTED** [review]
Wow, this is signifant work! Kudos. Regarding marking them unbroken, could you please squash that part into the commit that marks them broken in the first place? So they cancel each other out?

**Comment by XVilka** [comment]
This looks like a significant problem though cc @Heersin
```
[XX] db/rzil/arm32 emulateme_vfp
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc 's main
e asm.bits=32
aezi
o malloc://0x1000 0x40000
o malloc://0x10 0x50000
e io.cache=1
ar sp=0x41000
aezsu 0x10548 # after vsub
ar d7
aezsu 0x10554 # after vmul
ar d7
aezs # after vadd
ar d7
aezs # after vcvt
ar d16
aezsu 0x1058c # before printf
ar r2
ar r3
pxq 16 @ sp
' bins/elf/emulateme_vfp.arm32
-- stdout
--- expected
+++ actual
@@ -1,7 +1,7 @@
 d7 = 0x370637bd40acccbc
 d7 = 0x383531a640acccbc
 d7 = 0x40accd1740acccbc
-d16 = 0x401599a2e0000000
-r2 = 0xa0000000
-r3 = 0x40159999
-0x00040fd0  0x3ee0c6f7a0000000  0x401599a2e0000000   .......>.......@
+d16 = 0x0000000000000000
+r2 = 0x00000000
+r3 = 0x370637bd
+0x00040fd0  0x0000000000000000  0x0000000000000000   ................

-- stderr
--- expected
+++ actual
@@ -1,0 +1,2 @@
+ERROR: RzIL: invalid instruction or lifting not implemented at address 0x00010558
+ERROR: RzIL: invalid instruction or lifting not implemented at address 0x00010558
```

**Comment by wargio** [comment]
no, is likely missing some enums.

**Comment by XVilka** [comment]
```
librz/asm/meson.build:56:40: ERROR: String 'next' cannot be converted to int
```

File: test/db/analysis/ppc
Comment (XVilka) [inline test/db/analysis/ppc]: 3 broken PPC tests still left

File: test/db/analysis/x86_32
Comment (XVilka) [inline test/db/analysis/x86_32]: And this one is still broken

File: test/db/analysis/x86_64
Comment (XVilka) [inline test/db/analysis/x86_64]: And these are still broken

File: test/db/asm/x86_16
Comment (XVilka) [inline test/db/asm/x86_16]: Are these two unfixable?

File: test/db/asm/x86_16
Comment (XVilka) [inline test/db/asm/x86_16]: And this one?

File: test/db/asm/x86_32
Comment (XVilka) [inline test/db/asm/x86_32]: What about this one?

File: test/db/asm/x86_64
Comment (XVilka) [inline test/db/asm/x86_64]: These as well

File: test/db/esil/arm_64
Comment (XVilka) [inline test/db/esil/arm_64]: Don't forget this one please

File: test/db/formats/mach0/imports
Comment (XVilka) [inline test/db/formats/mach0/imports]: And this one

File: test/db/formats/mach0/objc
Comment (XVilka) [inline test/db/formats/mach0/objc]: And these

File: test/db/rzil/ppc32
Comment (XVilka) [inline test/db/rzil/ppc32]: This as well

File: test/db/tools/rz
Comment (XVilka) [inline test/db/tools/rz]: And this

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
@Rot127 once you have time, please look at the PPC changes, and we should test these on the capstone side

File: test/db/analysis/x86_32
Comment (wargio) [reply test/db/analysis/x86_32]: those 2 i don't even know if they are correct in the first place.

**Comment by XVilka** [comment]
Failed unit tests:
```
 17/117 rizin:unit / analysis_op                           FAIL            5.68s   killed by signal 11 SIGSEGV
>>> MALLOC_PERTURB_=222 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/build/test/unit/test_analysis_op

 91/117 rizin:unit / tokens                                FAIL            0.96s   killed by signal 11 SIGSEGV
>>> MALLOC_PERTURB_=80 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/build/test/unit/test_tokens
```

More build failures on capstone 3:
```
librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o -MF librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o.d -o librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o -c ../librz/analysis/arch/arm/arm_il32.c
../librz/analysis/arch/arm/arm_il32.c: In function ‘dt2fmt’:
../librz/analysis/arch/arm/arm_il32.c:201:14: error: ‘ARM_VECTORDATA_F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_U16’?
  201 |         case ARM_VECTORDATA_F16:
      |              ^~~~~~~~~~~~~~~~~~
      |              ARM_VECTORDATA_U16
../librz/analysis/arch/arm/arm_il32.c:201:14: note: each undeclared identifier is reported only once for each function it appears in
../librz/analysis/arch/arm/arm_il32.c: In function ‘cvt_isize’:
../librz/analysis/arch/arm/arm_il32.c:3512:14: error: ‘ARM_VECTORDATA_F16U32’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F16F32’?
 3512 |         case ARM_VECTORDATA_F16U32:
      |              ^~~~~~~~~~~~~~~~~~~~~
      |              ARM_VECTORDATA_F16F32
../librz/analysis/arch/arm/arm_il32.c:3515:14: error: ‘ARM_VECTORDATA_U32F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F32F16’?
 3515 |         case ARM_VECTORDATA_U32F16:
      |              ^~~~~~~~~~~~~~~~~~~~~
      |              ARM_VECTORDATA_F32F16
../librz/analysis/arch/arm/arm_il32.c:3526:14: error: ‘ARM_VECTORDATA_F16U16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F32U16’?
 3526 |         case ARM_VECTORDATA_F16U16:
      |              ^~~~~~~~~~~~~~~~~~~~~
      |              ARM_VECTORDATA_F32U16
../librz/analysis/arch/arm/arm_il32.c:3529:14: error: ‘ARM_VECTORDATA_U16F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_U16F32’?
 3529 |         case ARM_VECTORDATA_U16F16:
      |              ^~~~~~~~~~~~~~~~~~~~~
      |              ARM_VECTORDATA_U16F32
```

File: test/db/analysis/x86_32
Comment (XVilka) [reply test/db/analysis/x86_32]: @wargio I created this test, it's the same issue as long time ago: https://github.com/capstone-engine/capstone/issues/111

**Comment by Heersin** [comment]
> Failed unit tests:
> 
> ```
>  17/117 rizin:unit / analysis_op                           FAIL            5.68s   killed by signal 11 SIGSEGV
> >>> MALLOC_PERTURB_=222 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/build/test/unit/test_analysis_op
> 
>  91/117 rizin:unit / tokens                                FAIL            0.96s   killed by signal 11 SIGSEGV
> >>> MALLOC_PERTURB_=80 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-1287038090/rizinorg/rizin/build/test/unit/test_tokens
> ```
> 
> More build failures on capstone 3:
> 
> ```
> librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o -MF librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o.d -o librz/analysis/librz_analysis.so.0.6.0.p/arch_arm_arm_il32.c.o -c ../librz/analysis/arch/arm/arm_il32.c
> ../librz/analysis/arch/arm/arm_il32.c: In function ‘dt2fmt’:
> ../librz/analysis/arch/arm/arm_il32.c:201:14: error: ‘ARM_VECTORDATA_F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_U16’?
>   201 |         case ARM_VECTORDATA_F16:
>       |              ^~~~~~~~~~~~~~~~~~
>       |              ARM_VECTORDATA_U16
> ../librz/analysis/arch/arm/arm_il32.c:201:14: note: each undeclared identifier is reported only once for each function it appears in
> ../librz/analysis/arch/arm/arm_il32.c: In function ‘cvt_isize’:
> ../librz/analysis/arch/arm/arm_il32.c:3512:14: error: ‘ARM_VECTORDATA_F16U32’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F16F32’?
>  3512 |         case ARM_VECTORDATA_F16U32:
>       |              ^~~~~~~~~~~~~~~~~~~~~
>       |              ARM_VECTORDATA_F16F32
> ../librz/analysis/arch/arm/arm_il32.c:3515:14: error: ‘ARM_VECTORDATA_U32F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F32F16’?
>  3515 |         case ARM_VECTORDATA_U32F16:
>       |              ^~~~~~~~~~~~~~~~~~~~~
>       |              ARM_VECTORDATA_F32F16
> ../librz/analysis/arch/arm/arm_il32.c:3526:14: error: ‘ARM_VECTORDATA_F16U16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_F32U16’?
>  3526 |         case ARM_VECTORDATA_F16U16:
>       |              ^~~~~~~~~~~~~~~~~~~~~
>       |              ARM_VECTORDATA_F32U16
> ../librz/analysis/arch/arm/arm_il32.c:3529:14: error: ‘ARM_VECTORDATA_U16F16’ undeclared (first use in this function); did you mean ‘ARM_VECTORDATA_U16F32’?
>  3529 |         case ARM_VECTORDATA_U16F16:
>       |              ^~~~~~~~~~~~~~~~~~~~~
>       |              ARM_VECTORDATA_U16F32
> ```

not existed in v3 ?

**Comment by wargio** [comment]
> not existed in v3 ?

i added some stuff, and i forgot to add the compiler fences.

File: librz/analysis/arch/ppc/ppc_il.c
Comment (Rot127) [inline librz/analysis/arch/ppc/ppc_il.c:594]: If there is an issue open about this please link it.

File: librz/analysis/arch/ppc/ppc_il_ops.c
Comment (Rot127) [inline librz/analysis/arch/ppc/ppc_il_ops.c:259]: Is this whole `rA ? VARG() : NULL` thing new in `v5`? And is the assumption, in this case correct, that `rA` will be `NULL` if it is not a register (because the operand value in Capstone is a union).

File: librz/analysis/arch/ppc/ppc_il_ops.c
Comment (Rot127) [inline librz/analysis/arch/ppc/ppc_il_ops.c:438]: Please point out what bug and link an issue if possible. Just in case someone needs to know.

Comment (Rot127) [reply librz/analysis/arch/ppc/ppc_il_ops.c:438]: Also for the bug comments below as well please.

**Review by Rot127 — COMMENTED** [review]
Only looked at the PPC part but looks good to me. ~~Although we should probably fix those bugs you mention in CS and make a `v5.1`.~~ Actually no. Lets just push for `v6` as much as possible. Please add the comments about the bugs for easier comprehension nonetheless.

File: librz/analysis/arch/ppc/ppc_il_ops.c
Comment (wargio) [reply librz/analysis/arch/ppc/ppc_il_ops.c:259]: tbh i don't know if it is a bug or else.
From the disassembly prospective, i disassembled these using objdump and it reported the same instruction.
but essentially instead of a register you can have `zero` as base (thus rA is NULL)

**Comment by Rot127** [comment]
I can not check it currently. But, if I recall correctly, the register was not added to the operand list of `cs_insn` if it was zero (only in v5. In v6 it is always added). But, because of the missing operand, the disponent is treated as register here and can give a valid result, right?

Please add a test case, for an instruction with a base register set to 0 and a disponent of a range like `1-PPC_REG_ENDING`.

14 Jul 2023 05:12:17 Giovanni ***@***.***>:

**Comment by wargio** [comment]
> But, because of the missing operand, the disponent is treated as register here and can give a valid result, right?

According to the ISA, when `0` is actually `0`, not `R0`

**Comment by wargio** [comment]
This is a test for one of those `NULL` cases.
`e8000002` ppc32le `lwa r0, 0(0)`

**Comment by Rot127** [comment]
Sorry, I would test it myself, but do not have access to my laptop currently.

If the immediate is 0 it would return NULL (because `0 = REG_INVALID` so no reg name). Could you please add a test case like `
> lwa r0, 0(4)`
.

14 Jul 2023 10:10:11 Giovanni ***@***.***>:

**Comment by wargio** [comment]
`lwa r0, 0(4)` should be `lwa r0, 0(r4)` => `e8040002`
`e9000012` => `lwa r8, 0x10(0)`

image

https://www.ibm.com/docs/en/aix/7.1?topic=set-lwa-load-word-algebraic-instruction

**Comment by Rot127** [comment]
@wargio Checked it on my machine. I was confused, so ignore my comments above. But thanks for making the effort to go into it.

File: librz/asm/p/asm_x86_cs.c
Comment (XVilka) [inline librz/asm/p/asm_x86_cs.c:93]: Add a link to the capstone issue maybe? https://github.com/capstone-engine/capstone/issues/111

File: test/db/analysis/x86_64
Comment (XVilka) [inline test/db/analysis/x86_64:1395]: Nice, even new one unbroken test!

**Review by XVilka — CHANGES_REQUESTED** [review]
Please squash some commits so that we have cleaner history and commit marks any test as BROKEN. See my other comments too. Except that is LGTM. Some of these numerous hacks will likely be gone once auto-sync PRs are merged.

File: librz/analysis/arch/ppc/ppc_il_ops.c
Comment (XVilka) [reply librz/analysis/arch/ppc/ppc_il_ops.c:438]: @wargio and this one too

**Review by XVilka — APPROVED** [review]
LGTM once Woodpecker is fixed - it doesn't look like a caching issue. MacOS zeroes buffers on initialization, maybe it is the reason?

File: librz/analysis/arch/ppc/ppc_il.c
Comment (wargio) [reply librz/analysis/arch/ppc/ppc_il.c:594]: https://github.com/capstone-engine/capstone/issues/2087

**Comment by XVilka** [comment]
Opened an issue for this test here: https://github.com/rizinorg/rizin/issues/3658

**Pull request merged**
